src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1121,10 +1145,66 @@ private static ImmutableArray<StateForCase> RemoveEvaluation(ImmutableArray<Stat
+            BoundDagTemp input1 = test.Input;
+            BoundDagTemp input2 = other.Input;
+            if (!input1.Equals(input2))
+            {
+                while (true)
+                {
+                    switch (input1.Source, input2.Source)
+                    {
+                        case (BoundDagIndexerEvaluation s1, BoundDagIndexerEvaluation s2):
+                            (BoundDagTemp s1Input, BoundDagTemp s1LengthTemp, int s1Index) = GetCanonicalInput(s1);
+                            (BoundDagTemp s2Input, BoundDagTemp s2LengthTemp, int s2Index) = GetCanonicalInput(s2);
+                            Debug.Assert(s1LengthTemp.Syntax is ListPatternSyntax);
+                            Debug.Assert(s2LengthTemp.Syntax is ListPatternSyntax);
+                            if (s1Input.Equals(s2Input) &&
+                                // We don't want to pair two indices within the same pattern.
+                                s1LengthTemp.Syntax != s2LengthTemp.Syntax)
+                            {
+                                Debug.Assert(s1LengthTemp.Equals(s2LengthTemp));
+                                if (s1Index == s2Index)
+                                {
+                                    break;
+                                }
+                                if (s1Index < 0 != s2Index < 0)
+                                {
+                                    Debug.Assert(state.RemainingValues.ContainsKey(s1LengthTemp));
+                                    var lengthValues = (IValueSet<int>)state.RemainingValues[s1LengthTemp];
+                                    if (lengthValues.IsEmpty)
+                                    {
+                                        return;
+                                    }
+
+                                    // Compute the length value that would make these two indices point to the same element.
+                                    int lengthValue = s1Index < 0 ? s2Index - s1Index : s1Index - s2Index;
+                                    if (lengthValues.All(BinaryOperatorKind.Equal, lengthValue))
+                                    {
+                                        // If the length is known to be exact, the two can definitely point to the same element.
+                                        break;
+                                    }
+                                    if (lengthValues.Any(BinaryOperatorKind.Equal, lengthValue))
+                                    {
+                                        // Otherwise, we add a test to make the result conditional on the length value.
+                                        lengthTest = new Tests.One(new BoundDagValueTest(syntax, ConstantValue.Create(lengthValue), s1LengthTemp));
+                                        break;
+                                    }
+                                }
+                            }
+                            return;
```

Comment:
nit: consider commenting on the `return;` statements (tests are considered unrelated)